modules/math/shared/src/main/scala/gsp/math/skycalc/ImprovedSkyCalc.scala

```diff
@@ -0,0 +1,224 @@
+
+package gsp.math.skycalc
+
+import cats.Eq
+import cats.implicits._
+import gsp.math.Angle
+import gsp.math.Coordinates
+import java.time.Instant
+import java.time.ZonedDateTime
+
+/**
+  * Improved version of SkyCalc that supports lunar calculations. All instance stuff is here;
+  * the trait is exclusively static stuff.
+  * @author brighton, rnorris
+  */
+final class ImprovedSkyCalc(siteLatitude: Angle, siteLongitude: Angle, siteAlt: Int)
+    extends ImprovedSkyCalcMethods {
+
+  // Site parameters
+  private val hoursLongitude  = -siteLongitude.toDoubleDegrees / 15.0
+  private val degreesLatitude = siteLatitude.toDoubleDegrees
+  private val siteAltitude    = siteAlt.toDouble
+
+  // calculated results
+  private var altitude                             = .0
+  private var hourAngle                            = .0
+  private var azimuth                              = .0
+  private var parallacticAngle                     = .0
+  private var airmass                              = .0
+  private var lunarSkyBrightness: java.lang.Double = .0
+  private var lunarDistance                        = .0
+  private var lunarIlluminatedFraction             = .0
+  private var totalSkyBrightness                   = .0
+  private var lunarPhaseAngle                      = .0
+  private var sunAltitude                          = .0
+  private var lunarElevation                       = .0
+
+  // caching for calculate()
+  private var cachedCoordinates: Option[Coordinates] = none
+  private var cachedInstant: Option[Instant]         = none
+  private var cachedCalculateMoon: Option[Boolean]   = none
+
+  private implicit val eqInstant: Eq[Instant] = Eq.by(i => (i.getEpochSecond, i.getNano))
+
+  def calculate(
+    coords:        Coordinates,
+    instant:       Instant,
+    calculateMoon: Boolean
+  ): Unit = { // Early exit if the parameters haven't changed.
+    if (coords.some =!= cachedCoordinates ||
+        instant.some =!= cachedInstant ||
+        calculateMoon.some =!= cachedCalculateMoon
+    ) {
+      cachedCoordinates = coords.some
+      cachedInstant = instant.some
+      cachedCalculateMoon = calculateMoon.some
+      val dateTime = DateTime(instant)
+      val jdut     = new DoubleRef
+      val sid      = new DoubleRef
+      val curepoch = new DoubleRef
+      setup_time_place(dateTime, hoursLongitude, jdut, sid, curepoch)
+      val objra    = coords.ra.toAngle.toDoubleDegrees / 15
+      val objdec   = coords.dec.toAngle.toDoubleDegrees
+      val objepoch = 2000.0
+      getCircumstances(
+        objra,
+        objdec,
+        objepoch,
+        curepoch.d,
+        sid.d,
+        degreesLatitude,
+        jdut,
+        calculateMoon
+      )
+    }
+  }
+
+  private def getCircumstances(
+    objra:         Double,
+    objdec:        Double,
+    objepoch:      Double,
+    curep:         Double,
+    sid:           Double,
+    lat:           Double,
+    jdut:          DoubleRef,
```

Comment:
What is `DoubleRef`?

Reply:
It's a wrapper for a mutable double. It's used in a number of places for output values. It's a far cry from FP but I don't know if we want to rewrite the whole thing in FP style.

Reply:
I found it below in the code, I thought it was something external